### PR TITLE
Fix predication of blockBroadcast 

### LINKED ISSUE #869: Use of `blockBroadcast` depends on thread predication
https://github.com/csarofeen/pytorch/blob/15f2ae2dd45d5988366e7365159ec54c8566816f/torch/csrc/jit/codegen/cuda/predicate_compute.cpp#L219

Unless thread-predicated, broadcast axes with thread parallelization do not need to use `blockBroadcast`. See how `kir::BroadcastOp` is handled in codegen (https://github.com/csarofeen/pytorch/blob/20_12_3_devel/torch/csrc/jit/codegen/cuda/codegen.cpp#L561).

This bug causes the inline predicate to fail to be generated. One example can be found with the `FusionSimpleGemm` test:

```
__global__ void kernel1(Tensor<float, 2> T0, Tensor<float, 2> T1, Tensor<float, 2> T5) {
  alignas(4) extern __shared__ char array[];
  void* shared_mem = array;
  if (((((blockIdx.z * 4) + threadIdx.z) < T0.size[0]) && (((blockIdx.y * 4) + threadIdx.y) < T1.size[1]))) {
    T5[(((blockIdx.z * 4) + threadIdx.z) * (1 * T1.size[1])) + (((blockIdx.y * 4) + threadIdx.y) * 1)] = 0;
  }
  float T6[1];
  T6[0] = 0;
  #pragma unroll 1
  for(size_t ki150 = 0; ki150 < (ceilDiv(T0.size[1], 32)); ++ki150) {
    float T3[1];
    T3[0]
       = T1[(((ki150 * 32) + threadIdx.x) * T1.stride[0]) + (((blockIdx.y * 4) + threadIdx.y) * T1.stride[1])];
    float T2[1];
    T2[0]
       = T0[(((blockIdx.z * 4) + threadIdx.z) * T0.stride[0]) + (((ki150 * 32) + threadIdx.x) * T0.stride[1])];
...
```

Clearly, `T3` and `T2` are missing their predicates. Note that unswitch isn't used in the test.

### PR DESCRIPTION
**Pull request by naoyam**
- Moved ThreadPredicateMap from a local variable to a class member variable of `GpuLower`
- Added `ir_utils::hasBlockSync(Expr*, ThreadPrediateMap)`
- Removed `TensorDomain::hasBlockBroadcast` as it is not always really correct.

Compute-sanitizer errors are also gone with this PR. All of the C++ and Python tests pass without any sanitizer error.

This would be likely to conflict with #844, but hopefully not a big deal. (@rdspring1)

Fixes #869

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp [hunk 1]
```diff
@@ -272,6 +272,35 @@ kir::Bool* ThreadPredicateMap::getExpr(const TensorView* out_tv) const {
...
+  // reads from all threads in its CTA can be done with no parallel
+  // broadcast. Only one thread will write to shared memory followed
+  // by a proper _syncthreads.
+  const bool output_smem = tv->getMemoryType() == MemoryType::Shared;
+
+  for (auto id : iter_domains) {
+    if (!id->isBroadcast()) {
+      continue;
+    }
+    if (id->isBlockDim() || (!output_smem && id->isThreadDim())) {
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp:296]: Nice, I don't know that we support broadcast across grid.

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp:296]: We don't, but that error is caught elsewhere.

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp:296]: Sorry, at least not in this sense of grid broadcast. We can broadcast across grid from inputs, but not a dimension that was reduced in the kernel.

File: torch/csrc/jit/codegen/cuda/parallel_type_bitmap.h [hunk 2]
```diff
@@ -31,6 +31,9 @@ class ParallelTypeBitmap {
   bool operator[](size_t pos) const;
   std::map<ParallelType, bool> getMap() const;
 
+  bool hasTID() const;
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/parallel_type_bitmap.h:46]: Comment these please.

**Review by csarofeen — APPROVED** [review]
LGTM just need some comments.

**Pull request merged**